### PR TITLE
Add a new blog

### PR DESCRIPTION
**Pull request by husiyu**

### CODE REVIEW

File: tidb-cascades-planner.md [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+Transformation 是一个接口类型，用来定义一个逻辑变换规则。
+
+*   `GetPattern()` 方法获取这个变换规则所需要匹配的一个 Pattern。
+
+*   由于 Pattern 只能描述算子的类型，不能描述 LogicalPlan 内部的内容约束，因此通过 `Match()` 方法可以判断更细节的匹配条件。例如 Pattern 只能描述我们想要一个 Join 类型的算子，但是却没法描述这个 Join 应该是 InnerJoin 或者是 LeftOuterJoin，这类条件就需要在 `Match()` 中进行判断。
+
+*   `OnTransform()` 方法中定义了变换规则的具体内容，返回的内容分别是新的 GroupExpr，是否删除旧的 `GroupExpr`，是否删除旧的 Group 中所有的 `GroupExpr`。
+
+```golang
+// Transformation defines the interface for the transformation rules.
```
Comment (francis0407) [inline tidb-cascades-planner.md]: Remove this line.

File: tidb-cascades-planner.md [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+  GetPattern() *memo.Pattern
+ 
+  Match(expr *memo.ExprIter) bool
+  
+  OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error)
+}
+```
+
+
+下面我们以一个变换规则：`PushSelDownAggregation` 为例，具体介绍上面三个方法的使用方式。
```
Comment (francis0407) [inline tidb-cascades-planner.md]: Add a link https://github.com/pingcap/tidb/blob/6a5955750014f239a41362059ced6d8ab420f7b4/planner/cascades/transformation_rules.go#L394 for `PushSelDownAggregation`.

File: tidb-cascades-planner.md [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+这个规则匹配的 Pattern 是 `Selection -> Aggregation`，作用则是将这个 Selection 下推到 Aggregation 下面，例如 SQL: `select a, sum(b) from t group by a having a > 10 and max(c) > 10` 中，having 条件里的 `a > 10` 可以下推到 Aggregation 的下方。更具体地来说，只要 Selection 当中的一个 Expression 里的所有列都出现在 group by 的分组列时，我们就可以把这个 Expression 进行下推。
+
+参考下图：
+
+![](media/tidb-cascades-planner/5.png)
+
+1.  在 Group0 中的 Selection 匹配到了 Pattern `Selection -> Aggregation`。
+
+2.  执行了 `OnTransform()` 的转换，Selection 中的 `a > 10` 条件被下推到了新的 Aggregation 下方，并且保留的条件 `max(c) < 10` 成为了一个新的 Selection。
```
Comment (francis0407) [inline tidb-cascades-planner.md]: `max(c) < 10` -> `max(c) > 10`

File: tidb-cascades-planner.md [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+* 在生成 Implementation 之前会先对当前的 Group 调用`fillGroupStats()`来填充LogicalProperty 里的统计信息。
+
+* 最后就是调用 `implGroupExpr()` 来产生 Implementation 和递归调用 `implGroup()`来搜索子 Group 的过程。
+
+**3. EnforcerRule**
+
+上文中我们没有详细介绍 Enforcer 的概念，我们在这里补充。例如我们有这样一个 SQL：
+
+```sql
+select b, sum(c) over (partition by b) from t`
```
Comment (francis0407) [inline tidb-cascades-planner.md]: remove the last '`'

File: tidb-cascades-planner.md [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+```
+
+这是一个带有 Window Function 的查询，Window 中以 `b `列为分组，由于目前 Window 的实现是需要下层算子根据分组列有序，当没有可以使 `b `列有序的索引时，我们必须在 Window 算子下面强制添加一个 Sort 算子来满足 Window 算子向下传递的 PhysicalProperty。
```
Comment (francis0407) [inline tidb-cascades-planner.md]: `b ` -> `b`

**Comment by husiyu** [comment]
@YiniXu9506 PTAL

**Comment by xuechunL** [comment]
LGTM

**Pull request merged**